### ISSUE TITLE
How to get line by line start and end timestamps? 

### ISSUE THREAD

**Issue by youssefavx** [lowerquality/gentle#245]
I tried the Gentle standalone app and it is incredible! Very impressive, just the degree of precision is astounding. 

I exported a CSV and it was timestamps for every single word. What I'd love to do is somehow get the start and end timestamp for every new line in the transcript. 

I actually tried to parse this file but failed many times after several attempts. I tried creating an index of words, and an index of lines, and then adding the word indexes to the line indexes in a dictionary. I also tried iterating through the csv for every line to check if a word is in the line, but that failed too. I then tried reformatting the CSV into a text file with an index for every word like so ("The 0 question 1 is 2 that 3..."etc., and then finding the nearest substring match for every line in my paragraph but that also failed because it wouldn't get the complete paragraph. 

I would love if Gentle had a feature for us to be able to pick either "paragraph level" "sentence level" or even just line by line. 

No need though if this is too big of a job. I'd love any help whatsoever with this problem.

**Comment by strob** [comment]
You'll probably have better luck with the JSON output. You can use the offset timing to find newline characters in the transcript.

**Comment by youssefavx** [comment]
Do you mean the end and satart offset values here? In this section "tools" is the end of that first line, and "we'll" is the beginning of a new line. I'm not sure I understand how from just those numbers I could extract a line break if I understand you correctly.

```
{
      "alignedWord": "tools",
      "case": "success",
      "end": 29.099999999999998,
      "endOffset": 463,
      "phones": [
        {
          "duration": 0.1,
          "phone": "t_B"
        },
        {
          "duration": 0.17,
          "phone": "uw_I"
        },
        {
          "duration": 0.16,
          "phone": "l_I"
        },
        {
          "duration": 0.16,
          "phone": "z_E"
        }
      ],
      "start": 28.509999999999998,
      "startOffset": 458,
      "word": "tools"
    },
    {
      "alignedWord": "we'll",
      "case": "success",
      "end": 30.7,
      "endOffset": 474,
      "phones": [
        {
          "duration": 0.09,
          "phone": "w_B"
        },
        {
          "duration": 0.09,
          "phone": "ih_I"
        },
        {
          "duration": 0.23,
          "phone": "l_E"
        }
      ],
      "start": 30.29,
      "startOffset": 469,
      "word": "We'll"
    },
```

**Comment by youssefavx** [comment]
If you mean subtracting the difference between the end offset of one word (at the end of a line) and the end offset of another word (at the start of a new line) to find the highest differences, there may be instances where that matches a new line. 

For example, I subtracted: 

End offset for 'tools' is:  463 (End of first line)
End offset for 'We’ll': 474 (Start of second line)

Difference: 11

End offset for 'deeply': 984 (second line end)
End offset for 'You': 993 (third line start)

difference: 9

End offset for 'furs': 1536 
End offset for 'so': 1545

Difference: 9

However, you do find examples where the difference is also 9 but there is no line break. 

For example, inside of one sentence: "at stealing" 

End offset for 'at': 1517
End offset for 'stealing': 1526

Difference: 9

**Comment by youssefavx** [comment]
Or do you mean using the end offset to check whether '\n' comes after the word?

**Comment by youssefavx** [comment]
Never mind, I think I figured it out from that little clue you gave me, thanks so much!  I'd use the end offset and look 5 characters forward then check if '\n' is there, if so, then get the start time for the next word.

**Comment by jkovzels** [comment]
> Never mind, I think I figured it out from that little clue you gave me, thanks so much! I'd use the end offset and look 5 characters forward then check if '\n' is there, if so, then get the start time for the next word.

Do not look for 5 characters forward. Your new line characters will always fall in between `endOffset` of any particular word and `startOffset` of the next one. End & start offset are indices on the first and the last (well, +1) letters of aligned word in original transcription.  So if you do:
```
transcription.substring(words[i].endOffset, words[i+1].startOffset).contains("\n") 
```
you will know that line break happened in between this 2 words.

Code is grossly oversimplified, but in principle should give you a direction.

**Comment by youssefavx** [comment]
Thanks so much! I’ll try this out. I actually got my code to work but there’s one or 2 errors in the output and it was extremely difficult to work with. Are there any negative effects you foresee in looking 5 characters forward? I assume that would be more brittle and more likely to break in edge cases?

**Comment by jkovzels** [comment]
Yes, 5 chars is some artificial [magic number](https://en.wikipedia.org/wiki/Magic_number_%28programming%29).
Plus, you can fit more then one new line in 5 characters. 
Plus, you should account for different [new lines](https://en.wikipedia.org/wiki/Newline) out there.
Besides, as you noted, you need to ignore all `not-in-transcript` and `not-in-audio` tokens when you are looking for next words. 
Plus, handle end of array cases properly. All the regular coding stuff in other words :)

**Comment by youssefavx** [comment]
This is all very informative and I’m very appreciative. As a beginner, and someone who has explored a bit of behavioral science I tend to shy away from learning in the typical way of doing everything in its correct order, and tend to gravitate towards learning through what is most relevant to me at that moment in time. I love this way of learning and I actually find I remember tons more stuff this way. I’d have never remembered a simple for loop if I’d never found a tremendous use for it in something I really cared about. Ultimately  as a consequence of lack of good learning materials out there (I’m unaware of any [programmed instruction](https://en.wikipedia.org/wiki/Programmed_learning) courses on programming, and most tutorials are geared towards what the tutor cares about) (or rather lack of adequate  looking on my part), there will definitely be many gaps in my understanding along the way so again just wanted to say THANK YOU! for filling some of those gaps

**Comment by youssefavx** [comment]
Okay, I tried your suggestion and it ended up getting rid of some of those weird errors. 

In cases where I could not find a timestamp, I asked the program to search for the nearest start of end offset and then get the timestamp for that. I don't know if that's a good way of doing it and would love to know if you have any better ideas on how to do that.

**Comment by natelawrence** [comment]
Just a simple suggestion: 

It would be possible for you to:
- choose a unique unicode symbol which does not appear anywhere else in your transcript (something pictorial, perhaps)
- paste your transcript into the spreadsheet software of your choice, prior to submitting it to Gentle
- copy your special unicode symbol and paste it onto the beginning (or end - or one symbol for the beginning and another for the end, if you wish) of every line in the neighboring column.
- copy all columns (symbols + transcript lines) and paste them into a plain text editor
- select all and cut them from the plain-text editor and paste them into Gentle.

When Gentle is finished aligning and you're looking at either the CSV or the JSON output, you can search for your special character and snap directly to each instance to examine the neighboring/proceeding/preceding word.

For presentation purposes, after you've harvested your timecodes, search and replace the output csv/json/html file (specifically only the 'transcript' portion of it) and replace your unique 'beginning of line/end of line' character(s) with something invisible, like a 'space' character. (It is important to replace your special character with something that is the equal number of characters, in order to preserve the link between character-indices and timecodes.)

As you may imagine, this solution comes from a mind which does not think in code in the slightest, so apologies for the workaround nature of it.

Additionally, it is possible to convert Gentle output to a subtitle format. 

By default, each word is its own solitary caption-block, however a tool such as Aegisub will allow you to highlight a number of words (each plotted on the timeline as Gentle detected them) and then choose to concatenate them into a single block. That operation will automatically use the beginning of the first highlighted word and the end of the last highlighted word in order to determine the timing of the entire selection (in your case each line).

If you use unique 
- beginning-of-line, 
- end-of-line, and 
- desired line-break-mid-line 
symbols then merging selections into caption blocks becomes significantly easier (easier still if I could merely code it) and one need only search-and-replace your chosen mid-caption-block-line-break symbol with Aegisub's own new-line symbol.

Converting Gentle output into a subtitle format also allows you to correct typos in one's transcript without misaligning timing information to the character-indices of all words following your edit in the entire transcript.

Hopefully, this provokes some new ideas for you.
.nl
